apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1.IntegrationTests/KmsTest.cs

```diff
@@ -0,0 +1,247 @@
+            TestHelpers.ValidateData(bucketName, objectName, data, _fixture.Client, new DownloadObjectOptions { EncryptionKey = key });
+            TestHelpers.ValidateData(bucketName, objectName, data, client);
+        }
+
+        [SkippableFact]
+        public void UploadObject_BucketHasNoDefaultKmsKey_UploadWithClientDefaultCsek_ExplicitKmsKey()
+        {
+            Skip.If(SkipTests);
+
+            string bucketName = CreateBucketWithDefaultKmsKeyName(KeyName1);
+            string objectName = TestHelpers.GenerateName();
+            var data = TestHelpers.GenerateData(1024);
+            EncryptionKey key = EncryptionKey.Generate();
+            var client = StorageClient.Create(encryptionKey: key);
+
+            // Upload an object using a client that has an encryption key, but also specify a KMS key. The KMS key should be used,
+            // and the CSEK should be ignored.
+            Object obj = client.UploadObject(bucketName, objectName, null, data, new UploadObjectOptions { KmsKeyName = KeyName1 });
+            Assert.StartsWith(VersionedKeyName1Prefix, obj.KmsKeyName);
+
+            // We should be able to download it again, even with a default (no CSEK) client.
+            TestHelpers.ValidateData(bucketName, objectName, data);
+        }
+
+        [SkippableFact]
+        public void UploadObject_BucketHasDefaultKmsKey_UploadWithNoSpecifiedKey_DownloadWithClientWithCsek()
+        {
+            Skip.If(SkipTests);
+            var client = _fixture.Client;
+
+            string bucketName = CreateBucketWithDefaultKmsKeyName(KeyName1);
+            string objectName = TestHelpers.GenerateName();
+            var data = TestHelpers.GenerateData(1024);
+
+            // Upload an object without specifying anything, so it should be encrypted with key 1.
+            Object obj = client.UploadObject(bucketName, objectName, null, data);
+            Assert.StartsWith(VersionedKeyName1Prefix, obj.KmsKeyName);
+
+            // We should be able to download the file even if we use a client that supplies CSEK by default.
+            EncryptionKey key = EncryptionKey.Generate();
+            var encryptionKeyClient = StorageClient.Create(encryptionKey: key);
+            TestHelpers.ValidateData(bucketName, objectName, data, encryptionKeyClient);
+        }
+
+        [SkippableFact]
+        public void UploadObject_SpecifyBothCsekAndKmsKey()
+        {
+            Skip.If(SkipTests);
+            var client = _fixture.Client;
+
+            string objectName = TestHelpers.GenerateName();
+            var data = TestHelpers.GenerateData(1024);
+
+            // It's invalid to specify both an EncryptionKey (that's non-None) and a key name
+            var options = new UploadObjectOptions { EncryptionKey = EncryptionKey.Generate(), KmsKeyName = KeyName1 };
+            Assert.Throws<ArgumentException>(() => client.UploadObject(_fixture.InitiallyEmptyBucket, objectName, null, data, options));
```

Comment:
Worth checking that the thrown exception contains the correct error message, to make sure it's failing for the expected reason?

Reply:
Hmm. Maybe. I'll check for the presence of the two property names...